### PR TITLE
Add ViViT

### LINKED ISSUE #15666: Add Video Vision Transformer
# 🌟 New model addition

## Model description

The ViViT is a transformer based model similar to Vision Transformer but operating on videos. In their paper, the authors present 4 types of possible architecture, and 2 are considered successful (unfactorised and factorised encoder models). Paper available at https://arxiv.org/abs/2103.15691.

The original research uses flax, and mentions usage of weights from large pretrained image models (ViT). I currently have an implementation of model 1, analogous to existing ViTModel. I have also written a flax-to-pytorch weights conversion (to use weights from the original repo) and I have written the adaptation of ViT-based weights to ViViT weights (based on the methodology in the paper).

## Open source status

* [ ] the model implementation is available: Original implementation in Scenic (flax) https://github.com/google-research/scenic/tree/main/scenic/projects/vivit
* [ ] the model weights are available: Flax checkpoints and TF SavedModel present here https://github.com/google-research/scenic/blob/main/scenic/projects/vivit/README.md
* [ ] who are the authors: Anurag Arnab, Mostafa Dehghani, Georg Heigold, Chen Sun, Mario Lučić, Cordelia Schmid @MostafaDehghani @anuragarnab

### PR DESCRIPTION
**Pull request by jegork**
# What does this PR do?

Fixes #15666. Reopening #20441, as I have missed the comments provided by @amyeroberts so the issue was closed by the bot. 

Add Video Vision Transformer to transformers. This PR implements a spacetime version of the Video Vision Transformer from the original paper. 

I have provided the model weights here https://huggingface.co/jegormeister/vivit-b-16x2-kinetics400 
I will try to add Factorised Encoder version later on (these are the two versions that authors provide weight for).

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? https://github.com/huggingface/transformers/issues/15666
- [x] Did you make sure to update the documentation with your changes? I have added the documentation, but I have troubles testing it as I couldn't run the preview command of the doc-builder, so if someone has the possibility to run and check it, I will be really grateful!
- [x] Did you write any new necessary tests? WIP

## Who can review?

@amyeroberts provided the last suggestions to the closed PR, so I hope you can review this one. Thanks!

### CODE REVIEW

**Comment by jegork** [comment]
@amyeroberts thank you for your comments in the previous PR! I have addressed your suggestions and also added the image processor test. 

However, I ran `make style`, however the pipeline was failing at check_code_quality. Therefore I've updated the testing dependencies which lead to black updating. But! When I run `make style` again, it gives the following output:

```sh
reformatted transformers/examples/research_projects/deebert/src/modeling_highway_bert.py
reformatted transformers/examples/research_projects/movement-pruning/emmental/modeling_bert_masked.py
reformatted transformers/src/transformers/models/reformer/modeling_reformer.py
reformatted transformers/src/transformers/models/vivit/modeling_vivit.py
reformatted transformers/tests/models/vivit/test_image_processing_vivit.py
```

So it fixes not only the files from this PR, but also the already existing ones. Therefore I have a question: should I only pus reformatted files from this PR or all?

**Comment by amyeroberts** [comment]
@jegork The files listed e.g. `transformers/examples/research_projects/deebert/src/modeling_highway_bert.py` should have the most recent formatting applied and shouldn't need to be updated with this PR. 

Could you rebase on main, make sure the most recent formatting packages are installed using `pip install -e .[quality]` and try `make style` again?

File: tests/models/vivit/test_modeling_vivit.py [hunk 1]
```diff
@@ -0,0 +1,350 @@
+# coding=utf-8
+# Copyright 2022 The HuggingFace Inc. team. All rights reserved.
```
Comment (amyeroberts) [inline tests/models/vivit/test_modeling_vivit.py:2]:
```suggestion
# Copyright 2023 The HuggingFace Inc. team. All rights reserved.
```

**Review by amyeroberts — COMMENTED** [review]
Very nice! ❤️  Mostly small nit comments. Only necessary update before merging is removing the dependence on `decord` in the doc examples. 

Let me know if the additional diffs from `make style` are still appearing after updating styling packages and rebasing.

File: tests/models/vivit/test_image_processing_vivit.py [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+    def test_image_processor_from_dict_with_kwargs(self):
+        image_processor = self.image_processing_class.from_dict(self.image_processor_dict)
+        self.assertEqual(image_processor.size, {"shortest_edge": 18})
+        self.assertEqual(image_processor.crop_size, {"height": 18, "width": 18})
+
+        image_processor = self.image_processing_class.from_dict(self.image_processor_dict, size=42, crop_size=84)
+        self.assertEqual(image_processor.size, {"shortest_edge": 42})
+        self.assertEqual(image_processor.crop_size, {"height": 84, "width": 84})
+
+    def test_batch_feature(self):
```
[the remark this answers is not in the document]
Comment (jegork) [reply tests/models/vivit/test_image_processing_vivit.py:112]: This part was copied from the VideoMAE tests as you suggested, but I have added the decorator

**Comment by jegork** [comment]
@amyeroberts Thanks for your comments. I have addressed everything. However, I still get the same problems with `make style` 
I did `git fetch upstream`, then `git rebase upstream/main`, `pip install -e ".[quality]"` after which I ran `make style`

Which resulted in the following output at the end:
```sh
black examples tests src utils setup.py
Skipping .ipynb files as Jupyter dependencies are not installed.
You can fix this by running ``pip install "black[jupyter]"``
reformatted /Users/jegorkitskerkin/Documents/projects/transformers/examples/research_projects/deebert/src/modeling_highway_bert.py
reformatted /Users/jegorkitskerkin/Documents/projects/transformers/examples/research_projects/movement-pruning/emmental/modeling_bert_masked.py
reformatted /Users/jegorkitskerkin/Documents/projects/transformers/src/transformers/models/vivit/modeling_vivit.py
reformatted /Users/jegorkitskerkin/Documents/projects/transformers/src/transformers/models/reformer/modeling_reformer.py
reformatted /Users/jegorkitskerkin/Documents/projects/transformers/tests/models/vivit/test_image_processing_vivit.py

All done! ✨ 🍰 ✨
5 files reformatted, 2380 files left unchanged.
ruff examples tests src utils setup.py --fix
/Library/Developer/CommandLineTools/usr/bin/make autogenerate_code
running deps_table_update
updating src/transformers/dependency_versions_table.py
/Library/Developer/CommandLineTools/usr/bin/make extra_style_checks
python utils/custom_init_isort.py
python utils/sort_auto_mappings.py
doc-builder style src/transformers docs/source --max_len 119 --path_to_docs docs/source
Overwriting content of src/transformers/models/vivit/modeling_vivit.py.
Cleaned 1 files!
python utils/check_doc_toc.py --fix_and_overwrite
```

As you can see, the same unrelated-to-this-PR-files are getting formatted

File: src/transformers/models/vivit/configuration_vivit.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+            Number of hidden layers in the Transformer encoder.
+        num_attention_heads (`int`, *optional*, defaults to 12):
+            Number of attention heads for each attention layer in the Transformer encoder.
+        intermediate_size (`int`, *optional*, defaults to 3072):
+            Dimensionality of the "intermediate" (i.e., feed-forward) layer in the Transformer encoder.
+        hidden_act (`str` or `function`, *optional*, defaults to `"gelu_fast"`):
+            The non-linear activation function (function or string) in the encoder and pooler. If string, `"gelu"`,
+            `"relu"`, `"selu"`, `"gelu_fast"` and `"gelu_new"` are supported.
+        hidden_dropout_prob (`float`, *optional*, defaults to 0.0):
+            The dropout probabilitiy for all fully connected layers in the embeddings, encoder, and pooler.
```
Comment (amyeroberts) [inline src/transformers/models/vivit/configuration_vivit.py:62]:
```suggestion
            The dropout probability for all fully connected layers in the embeddings, encoder, and pooler.
```

File: src/transformers/models/vivit/modeling_vivit.py [hunk 4]
```diff
@@ -0,0 +1,734 @@
...
+        self.layernorm_after = nn.LayerNorm(config.hidden_size, eps=config.layer_norm_eps)
+
+    def forward(self, hidden_states, head_mask=None, output_attentions=False):
+        self_attention_outputs = self.attention(
+            self.layernorm_before(hidden_states),  # in Vivit, layernorm is applied before self-attention
+            head_mask,
+            output_attentions=output_attentions,
+        )
+        attention_output = self_attention_outputs[0]
+        outputs = self_attention_outputs[1:]  # add self attentions if we output attention weights
```
Comment (amyeroberts) [inline src/transformers/models/vivit/modeling_vivit.py:297]: small nit: putting the comments on the line above the relevant piece of code

```suggestion
        self_attention_outputs = self.attention(
            # in Vivit, layernorm is applied before self-attention
            self.layernorm_before(hidden_states),
            head_mask,
            output_attentions=output_attentions,
        )
        attention_output = self_attention_outputs[0]
        # add self attentions if we output attention weights
        outputs = self_attention_outputs[1:]
```

File: src/transformers/models/vivit/modeling_vivit.py [hunk 5]
```diff
@@ -0,0 +1,734 @@
...
+
+        >>> # video clip consists of 300 frames (10 seconds at 30 FPS)
+        >>> file_path = hf_hub_download(
+        ...     repo_id="nielsr/video-demo", filename="eating_spaghetti.mp4", repo_type="dataset"
+        ... )
+        >>> container = av.open(file_path)
+
+        >>> # sample 16 frames
+        >>> indices = sample_frame_indices(clip_len=32, frame_sample_rate=1, seg_len=len(videoreader))
+        >>> video = videoreader.get_batch(indices).asnumpy()
```
Comment (amyeroberts) [inline src/transformers/models/vivit/modeling_vivit.py:549]:
```suggestion
        >>> # sample 32 frames
        >>> indices = sample_frame_indices(clip_len=32, frame_sample_rate=1, seg_len=len(videoreader))
        >>> video = videoreader.get_batch(indices).asnumpy()
```

File: src/transformers/models/vivit/modeling_vivit.py [hunk 6]
```diff
@@ -0,0 +1,734 @@
...
```
Comment (amyeroberts) [inline src/transformers/models/vivit/modeling_vivit.py:685]:
```suggestion
        >>> # sample 32 frames
        >>> indices = sample_frame_indices(clip_len=32, frame_sample_rate=1, seg_len=len(videoreader))
        >>> video = videoreader.get_batch(indices).asnumpy()
```

File: src/transformers/models/vivit/modeling_vivit.py [hunk 7]
```diff
@@ -0,0 +1,734 @@
...
+
+        >>> feature_extractor = VivitImageProcessor.from_pretrained("google/vivit-b-16x2-kinetics400")
+        >>> model = VivitForVideoClassification.from_pretrained("google/vivit-b-16x2-kinetics400")
+
+        >>> inputs = feature_extractor(list(video), return_tensors="pt")
```
Comment (amyeroberts) [inline src/transformers/models/vivit/modeling_vivit.py:690]:
```suggestion
        >>> image_processor = VivitImageProcessor.from_pretrained("google/vivit-b-16x2-kinetics400")
        >>> model = VivitForVideoClassification.from_pretrained("google/vivit-b-16x2-kinetics400")

        >>> inputs = image_processor(list(video), return_tensors="pt")
```

File: src/transformers/models/vivit/image_processing_vivit.py [hunk 8]
```diff
@@ -0,0 +1,407 @@
...
+            `preprocess` method.
+        do_rescale (`bool`, *optional*, defaults to `True`):
+            Whether to rescale the image by the specified scale `rescale_factor`. Can be overridden by the `do_rescale`
+            parameter in the `preprocess` method.
+        rescale_factor (`int` or `float`, *optional*, defaults to `1/255`):
+            Defines the scale factor to use if rescaling the image. Can be overridden by the `rescale_factor` parameter
+            in the `preprocess` method.
+        do_zero_centering (`bool`, *optional*, defaults to `True`):
+            Whether to rescale the image to values between [-1 - 1] instead of [0 - 1]. Can be overriden by the
+            `do_zero_centering` in the `preprocess` method..
```
Comment (amyeroberts) [inline src/transformers/models/vivit/image_processing_vivit.py:93]:
```suggestion
            `do_zero_centering` in the `preprocess` method.
```

File: src/transformers/models/vivit/image_processing_vivit.py [hunk 9]
```diff
@@ -0,0 +1,407 @@
...
+        """
+        if offset:
+            rescaled_image = (image - 127.5) * scale
+            if data_format is not None:
+                rescaled_image = to_channel_dimension_format(rescaled_image, data_format)
+            rescaled_image = rescaled_image.astype(np.float32)
+        else:
+            rescaled_image = rescale(image, scale=scale, data_format=data_format, **kwargs)
+
+        return rescale(image, scale=scale, data_format=data_format, **kwargs)
```
Comment (amyeroberts) [inline src/transformers/models/vivit/image_processing_vivit.py:225]: This function doesn't shift the image value between [-1, 1] if offset is `True` it creates `rescaled_image` and the doesn't use it. 

```suggestion
        image = image.astype(np.float32)        
        if offset:
        	image = image - 127.5
        return rescale(image, scale=scale, data_format=data_format, **kwargs)
```

File: src/transformers/models/vivit/image_processing_vivit.py [hunk 10]
```diff
@@ -0,0 +1,407 @@
...
+    def rescale(
+        self,
+        image: np.ndarray,
+        scale: Union[int, float],
+        offset: bool = True,
+        data_format: Optional[Union[str, ChannelDimension]] = None,
+        **kwargs,
+    ):
+        """
+        Rescale an image by a scale factor. image = image * scale.
```
Comment (amyeroberts) [inline src/transformers/models/vivit/image_processing_vivit.py:205]:
```suggestion
                Rescale an image by a scale factor. 
        
        If offset is `True`, image scaled between [-1, 1]: image = (image - 127.5) * scale. 
        If offset is `False`, image scaled between [0, 1]: image = image * scale
```

File: src/transformers/models/vivit/convert_vivit_flax_to_pytorch.py [hunk 11]
```diff
@@ -0,0 +1,236 @@
...
+
+    model = VivitForVideoClassification(config)
+    model.load_state_dict(new_state)
+    model.eval()
+
+    extractor = get_processor()
+
+    video = prepare_video()
+    inputs = extractor(video, return_tensors="pt")
+    torch.save(inputs, "video_processed.pt")
```
Comment (amyeroberts) [inline src/transformers/models/vivit/convert_vivit_flax_to_pytorch.py:216]: Why save the inputs out?

**Review by amyeroberts — COMMENTED** [review]
Thanks for adding this model! 

Overall very good PR - just a few small nits here and there. Only thing to be addressed is the `rescale` logic in the image processor. 

I'm not sure why some other files have changed if you've rebased and have the most recent formatting settings. There's only a small number of differences and they actually improve the formatting, so let's leave them be.

File: src/transformers/models/vivit/convert_vivit_flax_to_pytorch.py (hunk 11 above)
Comment (jegork) [reply src/transformers/models/vivit/convert_vivit_flax_to_pytorch.py:216]: @amyeroberts my bad, forgot to remove it after debugging. Fixed

**Comment by jegork** [comment]
@amyeroberts yep, everything works and passes locally

**Comment by amyeroberts** [comment]
@jegork Thanks for confirming. I'm going to rerun CircleCI in case there was just some transient issue with the run. If it persists we can dig a bit more into it.

File: tests/models/vivit/test_modeling_vivit.py [hunk 12]
```diff
@@ -0,0 +1,350 @@
...
+        use_labels=True,
+        num_labels=10,
+        image_size=224,
+        num_frames=8,  # decreased, because default 32 takes too much RAM at inference
+        tubelet_size=[2, 16, 16],
+        num_channels=3,
+        hidden_size=768,
+        num_hidden_layers=12,
+        num_attention_heads=12,
+        intermediate_size=3072,
```
Comment (amyeroberts) [inline tests/models/vivit/test_modeling_vivit.py:61]: One thing to note here is that many of these values are very large and will cause the tests to run slowly. Could you update `image_size`, `num_frames`, `tubelete_size`, `hidden_size`, `num_attention_heads`, `num_hidden_layers` and `intermediate_size` to all be as small as possible? You can look at other model's test files to find typical values to use.

**Comment by adildahlan** [comment]
Thanks @amyeroberts and @jegork for working on this, we look forward to using the ViVit model!

File: tests/models/vivit/test_modeling_vivit.py (hunk 12 above)
Comment (jegork) [reply tests/models/vivit/test_modeling_vivit.py]: thanks! i've decreased those values and the tests pass now!

File: src/transformers/models/vivit/image_processing_vivit.py [hunk 13]
```diff
@@ -0,0 +1,404 @@
...
+        self,
+        do_resize: bool = True,
+        size: Dict[str, int] = None,
+        resample: PILImageResampling = PILImageResampling.BILINEAR,
+        do_center_crop: bool = True,
+        crop_size: Dict[str, int] = None,
+        do_rescale: bool = True,
+        rescale_factor: Union[int, float] = 1 / 255,
+        offset: bool = True,
+        do_normalize: bool = False,
```
Comment (amyeroberts) [inline src/transformers/models/vivit/image_processing_vivit.py:117]: Just double checking this should be `False`. I'm sure it's OK, but unusual

File: src/transformers/models/vivit/image_processing_vivit.py [hunk 14]
```diff
@@ -0,0 +1,404 @@
...
+    def rescale(
+        self,
+        image: np.ndarray,
+        scale: Union[int, float],
+        offset: bool = True,
+        data_format: Optional[Union[str, ChannelDimension]] = None,
+        **kwargs,
+    ):
+        """
+                Rescale an image by a scale factor.
```
Comment (amyeroberts) [inline src/transformers/models/vivit/image_processing_vivit.py:205]:
```suggestion
        Rescale an image by a scale factor.
```

File: src/transformers/models/vivit/image_processing_vivit.py [hunk 15]
```diff
@@ -0,0 +1,404 @@
...
+            scale (`int` or `float`):
+                Scale to apply to the image.
+           offset (`bool`, *optional*):
+                Whether to scale the image in both negative and positive directions.
+            data_format (`str` or `ChannelDimension`, *optional*):
+                The channel dimension format of the image. If not provided, it will be the same as the input image.
+        """
+        image = image.astype(np.float32)
+        if offset:
+            image = image - 127.5
```
Comment (amyeroberts) [inline src/transformers/models/vivit/image_processing_vivit.py:222]:
```suggestion
            image = image - (scale / 2)
```

File: src/transformers/models/vivit/image_processing_vivit.py [hunk 16]
```diff
@@ -0,0 +1,404 @@
...
+        return_tensors: Optional[Union[str, TensorType]] = None,
+        data_format: ChannelDimension = ChannelDimension.FIRST,
+        **kwargs,
+    ) -> PIL.Image.Image:
+        """
+        Preprocess an image or batch of images.
+
+        Args:
+            videos (`ImageInput`):
+                Image to preprocess.
```
Comment (amyeroberts) [inline src/transformers/models/vivit/image_processing_vivit.py:321]:
```suggestion
                Video frames to preprocess.
```

**Review by amyeroberts — APPROVED** [review]
Thanks for all the work adding this model! 

Just some nits to address and then we're good to merge!

File: src/transformers/models/vivit/image_processing_vivit.py (hunk 13 above)
Comment (jegork) [reply src/transformers/models/vivit/image_processing_vivit.py]: thanks for pointing it out! i've double checked the original code and it should've been `True`

**Comment by amyeroberts** [comment]
@jegork Is the PR OK to merge? Or are there any other commits you'd like to push before I press the big green button? 🟢

**Comment by jegork** [comment]
@amyeroberts I think it's ready to be merged. Thanks for your help!

**Comment by NielsRogge** [comment]
Hi @jegork congrats on your amazing contribution!

is it ok if we transfer the ViViT checkpoints to the `google` organization on the hub? (assuming they are officially released checkpoints by Google)

**Comment by jegork** [comment]
Hey @NielsRogge, thanks!

Sure

**Pull request merged**